README.md
# qcTimepicker

Simple jQuery dropdown timepicker. Small, accessible, and highly customizable.

[**Demo**](https://rinkattendant6.github.io/qcTimepicker/)

[**Documentation**](https://documentup.com/RinkAttendant6/qcTimepicker)

[![License MPL-2.0](https://img.shields.io/badge/license-MPL--2.0-yellowgreen.svg)](https://github.com/RinkAttendant6/qcTimepicker/blob/master/LICENSE)
[![Build Status](https://travis-ci.org/RinkAttendant6/qcTimepicker.svg?branch=master)](https://travis-ci.org/RinkAttendant6/qcTimepicker)

## Installation

### Bower (recommended)

    bower install qcTimepicker

### Manually

Download the files in the `dist/` directory.

## Usage

Load the plugin by including the script on the page:

```html
<script src='path/to/qcTimepicker.min.js' charset='utf-8'></script>
```

## Initialization

Initialize the plugin by calling the `.qcTimepicker()` function on a jQuery
object:

```js
$('.timepickers').qcTimepicker();
```

Please note that only `input` elements will be initialized.

It is strongly recommended that the input elements have the type `time`,
although all inputs that match the selector will be initialized.

Labels that are associated with the affected `input` element will be mapped to
the newly created dropdown element.

## Options

Options can be set during the initialization process by passing in an object
as a parameter:

```js
$('.timepickers').qcTimepicker({
    step: 900
});
```

### classes

**Type:** string or Array

**Default:** '' (empty string)

If specified, the class(es) will be applied to the each of the dropdowns
created by qcTimepicker.

### format

**Type:** string

**Default:** 'H:mm'

The format of the time as displayed in the dropdown.
[ICU symbols](http://userguide.icu-project.org/formatparse/datetime) are used
for the format. Supported symbols include: a, h, hh, H, HH, k, kk, K, KK, m,
mm, s, ss, A

To insert a literal symbol into the output string, escape the character with a
backslash. Currently, output strings cannot contain backslashes.

### minTime

**Type:** string or Date

**Default:** '00:00:00' (midnight)

The minimum time (lower-bound of range) in the dropdown. The value must be
specified as a string in 24-hour format or as a Date object. If a string is
given, less-significant units may be omitted if they are zero. Hours are
required.

Examples:

  - '11' => '11:00:00'
  - '11:00' => '11:00:00'
  - '11:30' => '11:30:00'

Leading zeros may be omitted:

  - '11:3' => '11:03:00'

Awkward times will be converted if possible:

  - '11:90:87' => '12:31:27'
  - '28:30:00' => '04:30:00'

Ignored if a `min` attribute is specified on the element of initialization.

**If the `min` attribute in the HTML is used, the value MUST be specified in
the official format: `HH:mm:ss` (Hours and minutes separated by colons,
optionally followed by a colon and number of seconds. Leading zeros are
required on all parts.)**

### maxTime

**Type:** string or Date

**Default:** '23:59:59'

The maximum time (upper-bound of range) in the dropdown. See `minTime` for
format.

Ignored if a `max` attribute is specified on the element of initialization.

**If the `max` attribute in the HTML is used, the value MUST be specified in
the official format: `HH:mm:ss` (Hours and minutes separated by colons,
optionally followed by a colon and number of seconds. Leading zeros are
required on all parts.)**

### step

**Type:** number

**Default:** 1800 (30 minutes)

The intervals of time displayed in the dropdown. The value represents the
number of seconds between each interval.

Ignored if a `step` attribute is specified on the element of initialization.

### placeholder

**Type:** string

**Default:** The text of an input element's `placeholder` attribute, if not present,
a hyphen will be displayed.

The text to display in the placeholder option. This is the first option
containing an empty value.

Ignored if a `data-placeholder` attribute or `placeholder` attribute is
specified on the element of initialization.

## Methods

### Toggle visibility

The visibility of the dropdown can be toggled by calling qcTimepicker with the
parameters `'show'` and `'hide'`, respectively.

```js
$('input').qcTimepicker('show');
$('input').qcTimepicker('hide');
```

### Destroy

To uninstantiate qcTimepicker and restore the original input, call qcTimepicker
with the `'destroy'` parameter:

```js
$('input').qcTimepicker('destroy');
```

### Step (increment and decrement)

To programatically increment or decrement the value in accordance with the
`step`, call qcTimepicker with the `stepUp` or `stepDown` parameters,
respectively.

This behaviour is consistent with the methods
[`stepUp`](https://www.w3.org/TR/html51/forms.html#dom-input-stepup) and 
[`stepDown`](https://http://www.w3.org/TR/html51/forms.html#dom-input-stepdown)
on the [`HTMLInputElement`](https://developer.mozilla.org/en/docs/Web/API/HTMLInputElement)
interface.

```html
<input value='08:00:00'>
```

```js
$('input').qcTimepicker({
    step: '02:00:00'
});

$('input').qcTimepicker('stepUp');
$('input').val() // returns 10:00:00

$('input').qcTimepicker('stepDown');
// Time is now back to 8:00
$('input').val() // returns 08:00:00
```

### Numeric and date values

qcTimepicker supports retrieving the time value as a number or a JavaScript
Date object. In the case of a number, the value returned is in milliseconds.

This behaviour is consistent with the properties
[`valueAsNumber`](https://www.w3.org/TR/html51/forms.html#dom-input-valueasnumber)
and [`valueAsDate`](https://www.w3.org/TR/html51/forms.html#dom-input-valueasdate)
on the [`HTMLInputElement`](https://developer.mozilla.org/en/docs/Web/API/HTMLInputElement)
interface.

```html
<input value='08:00:00'>
```

```js
$('input').qcTimepicker();

// As a number
var foo = $('input').qcTimepicker('valueAsNumber');
console.log(foo); // outputs 28800000

// As a Date
var bar = $('input').qcTimepicker('valueAsDate');

// The string output above differs based on your computer's current time zone
console.log(bar); // outputs Thu Jan 01 1970 03:00:00 GMT-0500 (Eastern Standard Time)

// Call getTime() or valueOf() for a primitive value
console.log(bar.getTime()); // outputs 28800000
```

CHANGELOG.md
# Changelog

## 1.2

- * Change build system from Grunt to Rollup
- * Change lint tool from JSHint to ESLint
- * Update development dependencies
- * Remove jQuery plugin metadata

## 1.1

 - + Add support for properties `valueAsDate` and `valueAsNumber
 - + Add support for methods `stepUp`, `stepDown`
 - + Add support for reading `min`, `max`, `step`, `disabled`, `readonly`, and
     `placeholder` attributes from `input` elements during initialization
 - * Change value of `step` option to number of seconds as per HTML5
     specification for the corresponding `step` attribute
 - * Code commenting annotation improvements
 - * DocumentUp support
 - + Copy of the unminified code added to build directory for proper source map
     linking

test/index.mjs
import { fileURLToPath } from 'node:url';
import path from 'node:path';

import {
    runQunitPuppeteer,
    printResultSummary,
    printFailedTests,
} from 'node-qunit-puppeteer';

try {
    const result = await runQunitPuppeteer({
        targetUrl: `file://${path.join(
            path.dirname(fileURLToPath(import.meta.url)),
            'index.html'
        )}`,
        redirectConsole: true,
    });

    printResultSummary(result, console);

    if (result.stats.failed > 0) {
        printFailedTests(result, console);
    }
} catch (ex) {
    console.error(ex);
}


test/jquery-loader.js
/*jslint browser: true, evil: true */
(function() {
    'use strict';
    var parts = document.location.search.slice(1).split('&'),
        length = parts.length,
        i,
        current,
        version = 'latest',
        file = '../lib/jquery/dist/jquery.min.js';

    for (i = 0; i < length; i += 1) {
        current = parts[ i ].split('=');
        if (current[ 0 ] === 'jquery') {
            version = current[ 1 ];
            break;
        }
    }

    if (version !== 'latest') {
        file = 'https://ajax.googleapis.com/ajax/libs/jquery/' + version + '/jquery.min.js';
    }

    document.write('<script src="' + file + '"></script>');
}());

Gruntfile.js
/*jslint node: true */
module.exports = function(grunt) {
    'use strict';

    require("matchdep").filterDev("grunt-*").forEach(grunt.loadNpmTasks);

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        uglify: {
            options: {
                banner: '/*! <%= pkg.name %> - v<%= pkg.version %> | (c) <%= grunt.template.today("yyyy") %> <%= pkg.author %>. | Licensed under the Mozilla Public License, version 2.0 */',
                sourceMap: true,
                preserveComments: 'some',
                mangle: {
                    except: ['jQuery']
                }
            },
            build: {
                files: {
                    'build/qcTimepicker.min.js': ['build/qcTimepicker.js']
                }
            }
        },
        copy: {
            main: {
                expand: true,
                cwd: 'src/',
                src: '*',
                dest: 'build/',
                flatten: true,
                filter: 'isFile'
            }
        }
    });

    grunt.registerTask('deploy', ['copy', 'uglify']);
};

rollup.config.js
import pkg from './package.json';
import clear from 'rollup-plugin-clear';
import copy from 'rollup-plugin-copy';
import { terser } from 'rollup-plugin-terser';

const header = `/*! ${pkg.name} - v${pkg.version} | (c) ${(new Date()).getFullYear()} ${pkg.author}
Licensed under the Mozilla Public License, version 2.0 */`;

export default {
    external: ['jquery'],
    input: 'src/qcTimepicker.js',
    output: {
        name: 'qcTimepicker',
        file: pkg.browser,
        format: 'iife',
        globals: {
            jquery: '$'
        },
        sourcemap: true
    },
    plugins: [
        clear({
            targets: ['dist']
        }),
        copy({
            targets: [
                {
                    src: 'src/qcTimepicker.js',
                    dest: 'dist/'
                }
            ]
        }),
        terser({
            format: {
                preamble: header
            }
        })
    ]
};


src/qcTimepicker.js
(function($) {
    'use strict';
    
    /**
     * Autoincrement number for inputs without identifiers
     * @type number
     */
    var i = 0;
    
    function validPartialTime(t) {
        return /^([0-1]?[0-9]|[2][0-3]):([0-5][0-9])(:([0-5][0-9]))?$/.test(t);
    }

    /**
     * Converts a partial time into seconds of day
     * @param {string} t A partial time as defined by RFC 3339
     * @param {boolean} [strict=false] Whether malformed partial times are accepted
     * @returns {number}
     */
    function timeToSeconds(t, strict) {
        var regex;
        
        strict = strict || false;
        regex = (strict ? /^([0-1]?[0-9]|[2][0-3]):([0-5][0-9])(:([0-5][0-9]))?$/ : /^\d+(:\d+){0,2}$/);
        
        if (!regex.test(t)) {
            throw 'InvalidArgumentException';
        }
        
        t = $.map(t.split(':'), function(part) {
            return parseInt(part, 10);
        });
        return ((t[0] * 3600 || 0) + (t[1] * 60 || 0) + (t[2] || 0)) % 86400;
    }
    
    /**
     * Converts a Date object into seconds of day
     * @param {!Date} date A Date object
     * @returns {number}
     */
    function dateInstanceToSeconds(date) {
        if(!(date instanceof Date)) {
            throw 'InvalidArgumentException';
        }
        
        return date.getHours() * 3600 + date.getMinutes() * 60 + date.getSeconds();
    }
    
    /**
     * Pads a number with leading zeros
     * @param {(number|string)} num The number to pad
     * @param {number} [len=2] The desired length
     * @returns {string} The padded number
     */
    function padNumber(num, len) {
        len = len || 2;
        num = num+'';
        
        if(num.length >= len) {
            return num;
        }
        
        var arr = [];
        arr.length = len + 1 - num.length;
        return arr.join('0') + num;
    }
    
    /**
     * Formats a time according to ICU formats
     * @param {string} format The format to use
     * @param {(number|string)} seconds The seconds of the day
     * @returns {string}
     */
    function formatTime(format, seconds) {
        seconds = Number(seconds);
        
        var time = {
            hours: Math.floor(seconds / 3600),
            minutes: Math.floor(seconds % 3600 / 60),
            seconds: seconds % 60
        },
        
        data = {
            'hh': padNumber((time.hours % 12) || 12),
            'h': ((time.hours % 12) || 12) + '',
            'HH': padNumber(time.hours),
            'H': time.hours + '',
            'kk': padNumber(time.hours + 1),
            'k': time.hours + 1 + '',
            'KK': padNumber(time.hours % 12 + 1),
            'K': time.hours % 12 + 1 + '',
            'mm': padNumber(time.minutes),
            'm': time.minutes + '',
            'ss': padNumber(time.seconds),
            's': time.seconds + '',
            'A': seconds * 1000 + '',
            'a': seconds < 43200 ? 'am' : 'pm'
        }, symbol, regexCallback;
        
        regexCallback = function($0, $1, $2) {
            return $1 ? $0 : data[$2];
        };
        
        for (symbol in data) {
            if (Object.prototype.hasOwnProperty.call(data, symbol)) {
                format = format.replace(new RegExp('(\\\\)?(' + symbol + ')', 'g'), regexCallback);
            }
        }
        
        // Strip slashes
        return format.replace(/\\/g, '');
    }
    
    /**
     * Sets the time of a dropdown
     * @param {string} time A partial-time as defined by RFC 3339
     * @param {!HTMLSelectElement} dropdown The timepicker element
     */
    function setTime(time, dropdown) {
        if($(dropdown.children).filter(function() {
            return this.value === time;
        }).length) {
            // Value exists in dropdown
            dropdown.value = time;
        }
    }
    
    /**
     * Methods available to qcTimepicker
     * @type Object.<string, function(...[object])>
     */
    var methods = {
        init: function(o) {
            var select = document.createElement('select'),
                placeholderOpt = document.createElement('option');

            placeholderOpt.value = '';
            select.appendChild(placeholderOpt);
            
            return this.filter('input').each(function() {
                var that = this,
                    tSelect = select.cloneNode(true),
                    opt,
                    labels = $('label[for="' + that.id + '"]'),
                    options, time;
                    
                options = $.extend({}, $.fn.qcTimepicker.defaults, o);
                
                // Prevent double-instantiation
                if(that.getAttribute('data-qctimepicker-id')) {
                    return;
                }
            
                // Add classes
                if(options.classes) {
                    if(typeof options.classes === 'object' && options.classes instanceof Array) {
                        $.each(options.classes, function(i, v) {
                            tSelect.className += ' ' + v;
                        });
                        tSelect.className = $.trim(tSelect.className);
                    } else if(typeof options.classes === 'string') {
                        tSelect.className = options.classes;
                    }
                }
                
                // Take into account max and min attributes where present
                if ((that.min || that.getAttribute('min')) && validPartialTime(that.min || that.getAttribute('min'))) {
                    options.minTime = timeToSeconds(that.min || that.getAttribute('min'), true);
                } else if (options.minTime instanceof Date) {
                    options.minTime = dateInstanceToSeconds(options.minTime);
                } else {
                    options.minTime = timeToSeconds(options.minTime);
                }
                
                if ((that.max || that.getAttribute('max')) && validPartialTime(that.max || that.getAttribute('max'))) {
                    options.maxTime = timeToSeconds(that.max || that.getAttribute('max'), true);
                } else if (options.maxTime instanceof Date) {
                    options.maxTime = dateInstanceToSeconds(options.maxTime);
                } else {
                    options.maxTime = timeToSeconds(options.maxTime);
                }

                // Take into account step attribute where present
                options.step = Number(that.step) ||
                               (that.getAttribute('step') ? Number(that.getAttribute('step')) : null) ||
                               (that.step === 'any' || that.getAttribute('step') === 'any' ? 1 : null) ||
                               options.step;
                
                if (typeof options.step !== 'number' || isNaN(options.step) || options.step <= 0 || options.step % 1 !== 0) {
                    options.step = $.fn.qcTimepicker.defaults.step;
                }

                // Generate options
                for(time = options.minTime; time <= options.maxTime; time += options.step) {
                    opt = document.createElement('option');
                    opt.innerHTML = formatTime(options.format, time);
                    opt.value = formatTime('HH:mm:ss', time);

                    tSelect.appendChild(opt);
                }
                
                // Copy over current value if possible
                setTime(that.value, tSelect);
                    
                // Keep controls in sync
                tSelect.onchange = function() {
                    that.value = this.value;
                };
                
                that.onchange = function() {
                    setTime(this.value, tSelect);
                };
                
                // If input is time, copy over properties as required
                if (that.type === 'time') {
                    that.step = options.step;
                    that.min = formatTime('HH:mm:ss', options.minTime);
                    that.max = formatTime('HH:mm:ss', options.maxTime);
                }
                
                // If input is required
                tSelect.required = (that.required || that.getAttribute('required') === 'required');
                
                // If input is disabled or readonly
                tSelect.disabled = that.readOnly || that.disabled;
                
                // Placeholder
                tSelect.firstChild.innerHTML = (that.dataset ? that.dataset.placeholder : false) || that.getAttribute('data-placeholder') || that.placeholder || that.getAttribute('placeholder') || options.placeholder;
                
                if(that.id) {
                    tSelect.id = that.id + '-qcTimepicker';
                } else {
                    tSelect.id = 'qcTimepicker-' + (i++);
                }
                
                // Change label references if necessary
                if(labels.length) {
                    $.each(labels, function(i, v) {
                        v.htmlFor = that.id + '-qcTimepicker';
                    });
                }
                
                // Append it!
                that.parentNode.insertBefore(tSelect, that.nextSibling);
                that.setAttribute('data-qctimepicker-id', tSelect.id);
                
                // Hide the input and make it non-focusable
                that._origDisplay = that.style.display;
                that._origTabIndex = that.tabIndex;
                
                that.style.display = 'none';
                that.tabIndex = -1;
            }).end();
        },
        
        hide: function() {
            return this.filter('input').each(function() {
                $('#' + this.getAttribute('data-qctimepicker-id')).hide();
            }).end();
        },
        
        show: function() {
            return this.filter('input').each(function() {
                $('#' + this.getAttribute('data-qctimepicker-id')).show();
            }).end();
        },
        
        destroy: function() {
            return this.filter('input[data-qctimepicker-id]').each(function() {
                var el = document.getElementById(this.getAttribute('data-qctimepicker-id'));
                el.parentNode.removeChild(el);
                
                this.style.display = this._origDisplay;
                this.tabIndex = this._origTabIndex;
                this.removeAttribute('data-qctimepicker-id');
            }).end();
        },
        
        options: function(o) {
            return this.filter('input[data-qctimepicker-id]').each(function() {
                var el = document.getElementById(this.getAttribute('data-qctimepicker-id'));
                
                if ('required' in o) {
                    el.required = Boolean(o.required);
                }
            }).end();
        },
        
        valueAsDate: function() {
            var input = this.filter('input[data-qctimepicker-id]')[0];
            if (input.value === '') {
                return null;
            }
            return input.valueAsDate || (new Date('Thu, 1 Jan 1970 ' + input.value + ' GMT'));
        },
        
        valueAsNumber: function() {
            var input = this.filter('input[data-qctimepicker-id]')[0];
            if (input.value === '') {
                return Number.NaN;
            }
            return input.type === 'time' ? input.valueAsNumber : Date.parse('Thu, 1 Jan 1970 ' + input.value + ' GMT');
        },
        
        stepUp: function() {
            return this.filter('input[data-qctimepicker-id]').each(function() {
                var el = document.getElementById(this.getAttribute('data-qctimepicker-id'));
                
                if (this.value === '' || el.selectedIndex === 0 || el.selectedIndex === el.children.length - 1) {
                    throw 'InvalidStateError';
                }
                
                el.selectedIndex += 1;
                $(el).trigger('change');
            }).end();
        },
        
        stepDown: function() {
            return this.filter('input[data-qctimepicker-id]').each(function() {
                var el = document.getElementById(this.getAttribute('data-qctimepicker-id'));
                
                if (this.value === '' || el.selectedIndex <= 1) {
                    throw 'InvalidStateError';
                }
                
                el.selectedIndex -= 1;
                $(el).trigger('change');
            }).end();
        }
    };
    
    /**
     * Calls to qcTimepicker
     * @param {(string|Object)=} args
     */
    $.fn.qcTimepicker = function(args) {
    
        if (methods[args]) {
            return methods[args].apply( this, Array.prototype.slice.call( arguments, 1 ));
        }
        
        if (typeof args === 'object' || !args) {
            // Default to "init"
            return methods.init.apply( this, arguments );
        }
        
        $.error('Method ' + args + ' does not exist on jQuery.qcTimepicker');
    };
    
    /**
     * Default options
     * @struct
     */
    $.fn.qcTimepicker.defaults = {
        classes: '',
        format: 'H:mm',
        minTime: '0:00:00',
        maxTime: '23:59:59',
        step: 1800,
        placeholder: '-'
    };
    
}(jQuery));